File: src/commands/remove.ts
import type { CommandDef } from "citty";
import { defineCommand } from "citty";
import { consola } from "consola";
import { vmsanPaths } from "../paths.ts";
import { VMService } from "../services/vm.ts";
import { createCommandLogger } from "../lib/logger/index.ts";

const removeCommand = defineCommand({
  meta: {
    name: "remove",
    description: "Remove one or more VMs (stops if running, then deletes)",
  },
  args: {
    vmIds: {
      type: "positional",
      description: "VM IDs to remove",
      required: true,
    },
    force: {
      type: "boolean",
      alias: "f",
      default: false,
      description: "Force removal of running VMs (stops them first)",
    },
  },
  async run({ args }) {
    const cmdLog = createCommandLogger("remove");
    const rawIds = [args.vmIds, ...args._];
    const vmIds = Array.from(new Set(rawIds.filter(Boolean)));

    if (vmIds.length === 0) {
      consola.error("No VM IDs provided.");
      cmdLog.emit();
      process.exitCode = 1;
      return;
    }

    const service = new VMService(vmsanPaths());

    // Validate all IDs exist upfront
    const missing: string[] = [];
    for (const id of vmIds) {
      if (!service.get(id)) missing.push(id);
    }
    if (missing.length > 0) {
      consola.error(`VM(s) not found: ${missing.join(", ")}`);
      cmdLog.emit();
      process.exitCode = 1;
      return;
    }

    // Block removal of non-stopped VMs unless --force is used
    if (!args.force) {
      const running: string[] = [];
      for (const id of vmIds) {
        const vm = service.get(id)!;
        if (vm.status !== "stopped") running.push(id);
      }
      if (running.length > 0) {
        consola.error(
          `Cannot remove running VM(s): ${running.join(", ")}. Stop them first or use --force (-f).`,
        );
        cmdLog.emit();
        process.exitCode = 1;
        return;
      }
    }

    // Remove sequentially
    const results: { vmId: string; success: boolean }[] = [];
    let hasErrors = false;
    for (const id of vmIds) {
      const log = consola.withTag(id);
      const vm = service.get(id)!;
      consola.debug(`VM ${id} status=${vm.status}, force=${args.force}`);

      log.start(`Removing ${id}...`);
      const result = await service.remove(id, { force: args.force });

      if (result.success) {
        log.success(`Removed ${id}`);
        results.push({ vmId: id, success: true });
      } else {
        consola.debug(`Removal failed for ${id}: ${result.error?.message ?? "unknown"}`);
        log.error(`Failed to remove ${id}: ${result.error?.message ?? "unknown error"}`);
        results.push({ vmId: id, success: false });
        hasErrors = true;
      }
    }

    cmdLog.set({ vmIds, results });
    if (hasErrors) {
      process.exitCode = 1;
    }
    cmdLog.emit();
  },
});

export default removeCommand as CommandDef;

File: hostd/internal/cli/remove.go
package cli

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	vmsanv1 "github.com/angelorc/vmsan/hostd/gen/vmsan/v1"
	"github.com/angelorc/vmsan/hostd/internal/gwclient"
	"github.com/angelorc/vmsan/hostd/internal/paths"
	"github.com/angelorc/vmsan/hostd/internal/vmstate"
	"github.com/spf13/cobra"
)

var removeCmd = &cobra.Command{
	Use:     "remove <vmId> [vmId...]",
	Aliases: []string{"rm"},
	Short:   "Remove one or more VMs (stops if running with --force)",
	Args:    cobra.MinimumNArgs(1),
	RunE:    runRemove,
}

func init() {
	removeCmd.Flags().BoolP("force", "f", false, "Force removal of running VMs (stops them first)")
	rootCmd.AddCommand(removeCmd)
}

func runRemove(cmd *cobra.Command, args []string) error {
	force, _ := cmd.Flags().GetBool("force")

	p := paths.Resolve()
	store := vmstate.NewStore(p.VmsDir)

	gw, err := gwclient.New()
	if err != nil {
		return fmt.Errorf("gateway connection failed: %w", err)
	}
	defer gw.Close()

	// Deduplicate
	seen := make(map[string]bool)
	var vmIDs []string
	for _, id := range args {
		if !seen[id] {
			seen[id] = true
			vmIDs = append(vmIDs, id)
		}
	}

	// Validate all IDs exist
	for _, id := range vmIDs {
		if _, err := store.Load(id); err != nil {
			return fmt.Errorf("VM %s not found", id)
		}
	}

	// Block removal of non-stopped VMs unless --force
	if !force {
		var running []string
		for _, id := range vmIDs {
			state, _ := store.Load(id)
			if state != nil && state.Status != "stopped" {
				running = append(running, id)
			}
		}
		if len(running) > 0 {
			return fmt.Errorf("cannot remove running VM(s): %v. Stop them first or use --force (-f)", running)
		}
	}

	type result struct {
		VmID    string `json:"vmId"`
		Success bool   `json:"success"`
		Message string `json:"message,omitempty"`
	}
	var results []result
	hasErrors := false

	for _, id := range vmIDs {
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		_, err := gw.DeleteVM(ctx, &vmsanv1.DeleteVMRequest{
			VmId:          id,
			Force:         force,
			JailerBaseDir: p.JailerBaseDir,
		})
		cancel()

		if err != nil {
			if !jsonOutput {
				fmt.Printf("Failed to remove %s: %v\n", id, err)
			}
			results = append(results, result{VmID: id, Success: false, Message: err.Error()})
			hasErrors = true
			continue
		}

		// Delete local state
		_ = store.Delete(id)

		if !jsonOutput {
			fmt.Printf("Removed %s\n", id)
		}
		results = append(results, result{VmID: id, Success: true})
	}

	if jsonOutput {
		data, _ := json.MarshalIndent(map[string]any{
			"results": results,
		}, "", "  ")
		fmt.Println(string(data))
	}

	if hasErrors {
		return fmt.Errorf("some VMs failed to remove")
	}
	return nil
}
